fix inference_mode with torch.compile

bdhirsh opened the pull request:
It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101219
* #100570

bdhirsh committed:
fix inference_mode with torch.compile

[ghstack-poisoned]

albanD commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -42,6 +42,10 @@ void FunctionalTensorWrapper::set_constructor_metadata() {
+  // E.g. when running torch.compile under inference mode, we need to make sure that
+  // for any inputs that were created outside of inference mode (so they are not inference tensors),
+  // then the functional wrappers that we wrap them with should also not be inference tensors.
+  version_counter_ = value_.unsafeGetTensorImpl()->version_counter();
```
Wouldn't this access to the version_counter raise an error on inference Tensors?

bdhirsh replied:
Talked offline - we're not accessing the version counter, just straight up copying the struct onto the wrapper.

Also - we copy the dispatch keyset from the inner tensor onto the wrapper, so if the inner tensor has the Autograd dispatch key (because it was created outside of inference mode), then the wrapper will as well (even though it was created in inference mode).

albanD reviewed (approved):
Sounds ok!

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

ezyang reviewed (approved):
Thank you for fixing this!!! What a pain haha

ezyang reviewed (approved):
Thank you for fixing this!!! What a pain haha

ezyang reviewed (approved):
Thank you for fixing this!!! What a pain haha

ezyang reviewed (approved):
Thank you for fixing this!!! What a pain haha

Chillee commented:
Wow Ed is really thankful for this fix.

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert -c "nosignal" -m "breaking inductor tests"
Add ciflow/inductor to run the tests on the pull request.

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh committed:
Update on "fix inference_mode with torch.compile"

It looks like inference_mode wasn't playing well with functionalization.

If you run torch.compile on a function, and the inputs to the function are tensors created outside of inference mode, then we need to make sure that when we created functional tensor wrappers for those inputs during compilation, those functional wrappers properly mirror whether or not the original tensor is an inference tensor.

Hopefully fixes https://github.com/pytorch/pytorch/issues/101151




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
